pubtools/_quay/container_image_pusher.py

```diff
@@ -0,0 +1,257 @@
+        source_ml = self._quay_client.get_manifest(source_ref, manifest_list=True)
+        simple_dest_refs = []
+        merge_mls_dest_refs = []
+
+        image_schema = "{host}/{namespace}/{repo}:{tag}"
+        namespace = self.target_settings["quay_namespace"]
+
+        for repo, tags in sorted(push_item.metadata["tags"].items()):
+            internal_repo = get_internal_container_repo_name(repo)
+            for tag in tags:
+                dest_ref = image_schema.format(
+                    host=self.quay_host,
+                    namespace=namespace,
+                    repo=internal_repo,
+                    tag=tag,
+                )
+                try:
+                    dest_ml = self._quay_client.get_manifest(dest_ref, manifest_list=True)
+                    LOG.info(
+                        "Getting missing archs between images '{0}' and '{1}'".format(
+                            source_ref, dest_ref
+                        )
+                    )
+                    missing_archs = ManifestListMerger.get_missing_architectures(source_ml, dest_ml)
+                    # Option 1: Destination doesn't contain extra archs, ML merging is unnecessary
+                    if not missing_archs:
+                        simple_dest_refs.append(dest_ref)
+                    # Option 2: Destination has extra archs, MLs will be merged
+                    else:
+                        merge_mls_dest_refs.append(dest_ref)
+                except requests.exceptions.HTTPError as e:
+                    # Option 3: Destination tag doesn't exist, no ML merging
+                    if e.response.status_code == 404:
+                        simple_dest_refs.append(dest_ref)
+                    else:
+                        raise e
+
+        if simple_dest_refs:
+            LOG.info(
+                "Copying image {0} to {1} destinations without merging manifest lists".format(
+                    source_ref, len(simple_dest_refs)
+                )
+            )
+            self.run_tag_images(source_ref, simple_dest_refs, True)
+        if merge_mls_dest_refs:
+            LOG.info(
+                "Copying image {0} to {1} destinations and merging manifest lists".format(
+                    source_ref, len(merge_mls_dest_refs)
+                )
+            )
+            self.run_merge_workflow(source_ref, merge_mls_dest_refs)
+
+    @log_step("Push images to Quay")
+    def push_container_images(self):
+        """
+        Push container images to Quay.
+
+        Two image types exist: source images and multiarch images. Non-source, single arch images
+        are not supported. In case of multiarch images, manifest list merging is performed if
+        destination image contains more architectures than source.
+        """
+        for item in self.push_items:
+            try:
+                source_ml = self._quay_client.get_manifest(
+                    item.metadata["pull_url"], manifest_list=True
```

Comment:
I wondered the same thing, I think I left a TODO comment somewhere about this. @midnightercz do you know?

Reply:
Do pull_url points to registry-proxy, so we shouldn't need any account for that (we don't have one for current implementation for pulp). But could we capture this idea to a new jira? It could be used to improve performance as I think quay.io should perform faster then internal read-only proxy

Reply:
It doesn't require credential to pull image from registry-proxy but does for get_manifest.

Reply:
I filed a Jira for investigating this in greater detail